src/libraries/System.Private.CoreLib/src/System/Runtime/InteropServices/GCHandle.T.cs

```diff
@@ -0,0 +1,100 @@
+    {
+        // The actual integer handle value that the EE uses internally.
+        private IntPtr _handle;
+
+        /// <summary>
+        /// Allocates a handle for the specified object.
+        /// </summary>
+        /// <param name="target">The object that uses the <see cref="GCHandle{T}"/>.</param>
+        public GCHandle(T target)
+        {
+            _handle = GCHandle.InternalAlloc(target, GCHandleType.Normal);
+        }
+
+        private GCHandle(IntPtr handle) => _handle = handle;
+
+        /// <summary>Determine whether this handle has been allocated or not.</summary>
+        public readonly bool IsAllocated => _handle != IntPtr.Zero;
+
+        /// <summary>Gets or sets the object this handle represents.</summary>
+        public readonly T Target
+        {
+            get
+            {
+                IntPtr handle = _handle;
+                GCHandle.ThrowIfInvalid(_handle);
+
+                // Skip the type check to provide maximum unsafety.
+                return Unsafe.As<T>(GCHandle.InternalGet(handle));
+            }
+            set
+            {
+                IntPtr handle = _handle;
+                GCHandle.ThrowIfInvalid(handle);
+
+                GCHandle.InternalSet(handle, value);
+            }
+        }
+
+        /// <summary>
+        /// Returns a new <see cref="GCHandle{T}"/> object created from a handle to a managed object.
+        /// </summary>
+        /// <param name="value">An <see cref="IntPtr"/> handle to a managed object to create a <see cref="GCHandle{T}"/> object from.</param>
+        /// <returns>A new <see cref="GCHandle{T}"/> object that corresponds to the value parameter.</returns>
+        /// <remarks>
+        /// The <see cref="IntPtr"/> representation of <see cref="GCHandle{T}"/> is not
+        /// interchangable with <see cref="GCHandle"/>.
+        /// </remarks>
+        public static GCHandle<T> FromIntPtr(IntPtr value)
+        {
+            GCHandle.ThrowIfInvalid(value);
+            return new GCHandle<T>(value);
+        }
+
+        /// <summary>
+        /// Returns the internal integer representation of a <see cref="GCHandle{T}"/> object.
+        /// </summary>
+        /// <param name="value">A <see cref="GCHandle{T}"/> object to retrieve an internal integer representation from.</param>
+        /// <returns>An <see cref="IntPtr"/> object that represents a <see cref="GCHandle{T}"/> object.</returns>
+        /// <remarks>
+        /// The <see cref="IntPtr"/> representation of <see cref="GCHandle{T}"/> is not
+        /// interchangable with <see cref="GCHandle"/>.
+        /// </remarks>
+        public static IntPtr ToIntPtr(GCHandle<T> value) => value._handle;
+
+        /// <summary>
+        /// Releases this <see cref="GCHandle{T}"/>.
+        /// </summary>
+        public void Dispose()
+        {
+            // Free the handle if it hasn't already been freed.
+            IntPtr handle = Interlocked.Exchange(ref _handle, IntPtr.Zero);
```

Comment:
This should not be an interlocked operation. Using this concurrently is not supported. It was one of the legacy behaviors from `GCHandle` that we wanted to get rid of with these new types. Same here and with the other two, of course 🙂

Reply:
Sure, I didn't tracked the discussion thread too much and was copying from `GCHandle`. I agree that every check should be skipped as long as we are able to throw exceptions.